src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs

```diff
@@ -263,6 +263,38 @@ private void PublishCode()
             _methodCodeNode.InitializeDebugVarInfos(_debugVarInfos);
 #if READYTORUN
             _methodCodeNode.InitializeInliningInfo(_inlinedMethods.ToArray());
+
+            // Detect cases where the instruction set support used is a superset of the baseline instruction set specification
```

Comment:
Is this saying it will detect `AVX` as a superset of `SSE2` or it only detects instruction sets that are a superset and share the same encoding?

Reply:
Its more subtle than that.

The default baseline instruction set is Sse, Sse2. However, the jit compiler is told to generate code assuming that it may use Sse, Sse2, Sse3, Ssse3, Sse41, Sse42, Lzcnt, Popcnt, and Pclmqdq. But if the compiler actually uses any of Sse3, Ssse3, Sse41, Sse42, Lzcnt, Popcnt, or Pclmqdq it will notify the crossgen2 managed code. That notification feeds into this logic and if so, we will then mark the method such that the generated code can only be used if the actual hardware supports the instruction set. That's all well and good, and considering that just about all hardware does support these instruction sets, we get crossgen R2R to generate ideal code for all these cases.

The less optimal case is where code attempts to use an Avx, or Avx2 feature, or such. In thoses cases, we will generate code as if the processor does NOT support the feature. In corelib, the idea is that we will audit our code reviews such that such a decision does not make the logic wrong (just slow) (This code auditing is already necessary for our current crossgen1 intrinsics support for corelib), but in the general case we cannot assume that the code will behave correctly if it runs on hardware that does support Avx, Avx2, etc. In that case, what this logic does is record the decision that Avx use was attempted and forbidden. Then the generated code will not be used at runtime, and the JIT will be invoked, if the hardware actually supports Avx/Avx2.

In addition, the --instruction-set argument allows adjusting the baseline instruction set. For instance, if the compiler is run with --instruction-set Sse42, --instruction-set Popcnt --instruction-set Lzcnt, --instruction-set Pclmqdq, then this logic will never kick in, and do anything, as the generated code will never exceed that of the baseline.

Reply:
Before this is merged, I'd really like to see copious documentation of the behavior. It is potentially quite confusing, and we need to be sure that the precise behavior is discoverable and understandable. Also, we should make sure that the JITDUMP indicates the information it has communicated to the VM, and that the log files at execution time include information about when and why methods are rejected for reasons such as this.

Reply:
Agreed that this needs documentation somewhere, the existing scheme is very nearly undescribed, and this is a substantial difference (and when this feature is complete, I don't expect to change the crossgen1 model, and this more advanced handling of intrinsics and such will remain restricted to crossgen2. I expect I'll put a document in the BoTR about this.

Adding JITDUMP data should be easy.

For logging, what sort of thing are you looking for? That's a feature all by itself, as we don't currently have any reasonable form of logging around R2R load behavior for functions. I could add an event, or possibly a STRESSLOG entry or an addition the the debug only LOG. @CarolEidt @tannergooding I'd like some feedback here.

Reply:
There's a method `SetReadyToRunRejectedPrecompiledCode` on `PrepareCodeConfig` that we already use to note when the precompiled R2R version of a method is not usable because of changes in dependent assemblies. This makes it into the ETL so jit event clients can see how many precompiled methods ended up not being usable.

Presumably you could extend this concept.

Reply:
Ah, that's convenient. I wasn't aware of that bit, but it should be fairly straightforward to add one bit that indicates that it was dropped due to an instruction set mismatch, and from there looking at the r2rdump output to figure out which exact instruction set was at fault should be straightforward.

Reply:
Also, the `SetReadyToRunRejectedPrecompiledCode` bit would be set for this condition as well. That bit is not tied to changes in dependent assemblies, its tied to changes in type layout, which are most likely triggered by changes in dependent assemblies, but there are other ways it could be different.